R/uncount.R

```diff
@@ -23,15 +24,18 @@
 uncount <- function(data, weights, .remove = TRUE, .id = NULL) {
   weights_quo <- enquo(weights)
   w <- dplyr::pull(dplyr::mutate(data, `_weight` = !! weights_quo))
+  w <- vec_cast(w, integer(), x_arg = "weights")
 
-  if (!is.numeric(w)) {
-    stop("`weights` must evaluate to a numeric vector", call. = FALSE)
+  if (any(w < 0)) {
+    abort("all elements of `weights` must be >= 0")
   }
 
-  out <- data[rep(seq_nrow(data), w), , drop = FALSE]
+  out <- vec_slice(data, vec_rep_each(vec_seq_along(data), w))
 
-  if (.remove && quo_is_symbol(weights_quo)) {
-    out[[as_string(get_expr(weights_quo))]] <- NULL
+  weights_is_col <- quo_is_symbol(weights_quo) &&
+    quo_name(weights_quo) %in% colnames(data)
+  if (.remove && weights_is_col) {
+    out <- dplyr::select(out, -!!as_name(weights_quo))
```

Comment:
Reasoning through what `as_name(quo)` is tricky — can you please keep the previous stringification? And then do `-all_of(weights_chr)` or similar?

Reply:
Good point. I just realised that it is much easier to use `any_of()` as I don't need `quo_name(weights_quo) %in% colnames(data)` then.